synthesizer/process/src/lib.rs

```diff
@@ -123,28 +154,103 @@ impl<N: Network> Process<N> {
-        // Add the stack to the process.
-        self.stacks.insert(*stack.program_id(), Arc::new(stack));
+    pub fn add_stack(&self, stack: Arc<Stack<N>>) -> Result<()> {
+        // Collect all direct and indirect external stacks.
+        let programs_to_add = stack.all_external_stacks();
+        // Obtain the lock on the stacks.
+        let mut stacks = self.stacks.lock();
+        // Determine which stacks still need to be added into the process.
+        let programs_to_add = programs_to_add
+            .into_iter()
+            .chain(std::iter::once((*stack.program_id(), stack))) // add the root stack.
+            .unique_by(|(id, _)|*id) // don't add duplicates.
+            .filter(|(program_id, _)| program_id != &ProgramID::<N>::from_str("credits.aleo").unwrap()) // don't add the credits.aleo stack.
+            .filter(|(program_id, _)| !stacks.contains(program_id)) // don't add stacks present in the cache.
+            .collect::<Vec<_>>();
+        // Determine the required capacity.
+        let current_capacity = stacks.cap().get().saturating_sub(stacks.len());
+        let mut required_capacity = programs_to_add.len().saturating_sub(current_capacity);
+        if required_capacity > 0 {
+            debug!("Evicting {required_capacity} stacks from the cache.");
+        }
+        // If the new stacks require more capacity, attempt to remove the least recently used stacks.
+        while required_capacity > 0 {
+            // To avoid dangling stacks, we only remove stacks that do not have more than 1 reference (e.g. as external stack).
+            // For this not to loop endlessly, we assume to always quickly find stacks with just one reference, so:
+            // 1. queried stacks must be short-lived, which is essential anyway to avoid memory leaks.
+            // 2. queried stacks must expire before a querying thread queries the next stack.
+            let root_program_ids = stacks
+                .iter()
+                .rev()
+                .filter_map(|(program_id, stack)| (std::sync::Arc::<_>::strong_count(stack) == 1).then_some(program_id))
+                .take(required_capacity)
+                .cloned()
+                .collect::<Vec<_>>();
+            // Lower the required capacity.
+            required_capacity = required_capacity.saturating_sub(root_program_ids.len());
+            // Evict the old stacks from the cache.
+            for program_id in root_program_ids {
+                if stacks.pop(&program_id).is_none() {
+                    bail!("Could not find expected program id in cache.")
+                }
+            }
+        }
+        // Add a new stacks into the cache.
+        for (program_id, stack) in programs_to_add {
+            stacks.put(program_id, stack);
+        }
+        Ok(())
+    }
+
+    /// Returns the size of the cache.
+    #[inline]
+    pub fn num_stacks(&self) -> usize {
+        self.stacks.lock().len()
```

Comment:
depending on the frequency of calls to `num_stacks`, we might actually want to use an `RwLock<LruCache>` this time and use `read` here instead

Reply:
Only using it in tests, will try to conditionally compile

Reply:
f1022e3e8